staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/compilation_test.go

```diff
@@ -658,3 +663,857 @@ func TestCelCompilation(t *testing.T) {
+			expectedCalcCost: 3145727,
+			setMaxLength:     100,
+			expectedSetCost:  302,
+		},
+		{
+			name:             "array of objects with all",
+			schemaGenerator:  genObjectArrayWithRule("self.all(x, true)"),
+			expectedCalcCost: 555128,
+			setMaxLength:     50,
+			expectedSetCost:  152,
+		},
+		{
+			name:             "map of numbers with all",
+			schemaGenerator:  genMapWithRule("number", "self.all(x, true)"),
+			expectedCalcCost: 1348169,
+			setMaxLength:     10,
+			expectedSetCost:  32,
+		},
+		{
+			name:             "map of numbers with has",
+			schemaGenerator:  genMapWithRule("number", "has(self.x)"),
+			expectedCalcCost: 0,
+			setMaxLength:     100,
+			expectedSetCost:  0,
+		},
+		{
+			name:             "map of strings with all",
+			schemaGenerator:  genMapWithRule("string", "self.all(x, true)"),
+			expectedCalcCost: 1179647,
+			setMaxLength:     3,
+			expectedSetCost:  11,
+		},
+		{
+			name:             "map of strings with has",
+			schemaGenerator:  genMapWithRule("string", "has(self.x)"),
+			expectedCalcCost: 0,
+			setMaxLength:     550,
+			expectedSetCost:  0,
+		},
+		{
+			name:             "map of booleans with all",
+			schemaGenerator:  genMapWithRule("boolean", "self.all(x, true)"),
+			expectedCalcCost: 943718,
+			setMaxLength:     100,
+			expectedSetCost:  302,
+		},
+		{
+			name:             "map of booleans with has",
+			schemaGenerator:  genMapWithRule("boolean", "has(self.x)"),
+			expectedCalcCost: 0,
+			setMaxLength:     1024,
+			expectedSetCost:  0,
+		},
+		{
+			name:             "string with contains",
+			schemaGenerator:  genStringWithRule("self.contains('test')"),
+			expectedCalcCost: 314574,
+			setMaxLength:     10,
+			expectedSetCost:  5,
+		},
+		{
+			name:             "string with startsWith",
+			schemaGenerator:  genStringWithRule("self.startsWith('test')"),
+			expectedCalcCost: 2,
+			setMaxLength:     15,
+			expectedSetCost:  2,
+		},
+		{
+			name:             "string with endsWith",
+			schemaGenerator:  genStringWithRule("self.endsWith('test')"),
+			expectedCalcCost: 2,
+			setMaxLength:     30,
+			expectedSetCost:  2,
+		},
+		{
+			name:             "concat string",
+			schemaGenerator:  genStringWithRule(`size(self + "hello") > size("hello")`),
+			expectedCalcCost: 314578,
+			setMaxLength:     4,
+			expectedSetCost:  7,
+		},
+		{
+			name:             "index of array with numbers",
+			schemaGenerator:  genArrayWithRule("number", "self[1] == 0.0"),
+			expectedCalcCost: 2,
+			setMaxLength:     5000,
+			expectedSetCost:  2,
+		},
+		{
+			name:             "index of array with strings",
+			schemaGenerator:  genArrayWithRule("string", "self[1] == self[1]"),
+			expectedCalcCost: 314577,
+			setMaxLength:     8,
+			expectedSetCost:  314577,
+		},
+		{
+			name:                       "O(n^2) loop with numbers",
+			schemaGenerator:            genArrayWithRule("number", "self.all(x, self.all(y, true))"),
+			expectedCalcCost:           9895601504256,
```

Comment:
okay, so O(n) ops are in the millions of cost units and O(n) are in the trillions of cost units. So our earlier thoughts to have the limit be set in the low billions still sounds reasonable.